File: backends/java/zappy-domain/src/test/java/com/zappymart/domain/builders/CartBuilder.java
package com.zappymart.domain.builders;

import com.zappymart.domain.cart.Cart;
import com.zappymart.domain.cart.CartLine;
import com.zappymart.domain.catalogue.Product;
import com.zappymart.domain.promotions.PromotionCode;
import com.zappymart.domain.promotions.PromotionRule;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

public final class CartBuilder {

    public static final Instant A_MOMENT = Instant.parse("2026-09-09T12:00:00Z");

    private final List<CartLine> lines = new ArrayList<>();
    private String id = "cart-01";
    private String customerId;
    private PromotionCode promotionCode;
    private PromotionRule promotionRule;

    public static CartBuilder aCart() {
        return new CartBuilder();
    }

    public CartBuilder withId(String newId) {
        this.id = newId;
        return this;
    }

    public CartBuilder ownedBy(String newCustomerId) {
        this.customerId = newCustomerId;
        return this;
    }

    public CartBuilder holding(Product product, int quantity) {
        lines.add(new CartLine("line-" + (lines.size() + 1), product, quantity));
        return this;
    }

    public CartBuilder withPromotion(String code, PromotionRule rule) {
        this.promotionCode = new PromotionCode(code);
        this.promotionRule = rule;
        return this;
    }

    public Cart build() {
        return Cart.of(id, customerId, lines, promotionCode, promotionRule, A_MOMENT);
    }
}

File: backends/kotlin/zappy-domain/src/testFixtures/kotlin/nl/zappymart/domain/builders/CartBuilder.kt
package nl.zappymart.domain.builders

import java.time.Instant
import nl.zappymart.domain.cart.Cart
import nl.zappymart.domain.cart.CartLine
import nl.zappymart.domain.catalogue.Product
import nl.zappymart.domain.promotions.AppliedPromotion
import nl.zappymart.domain.promotions.PromotionKind
import nl.zappymart.domain.shared.Money
import nl.zappymart.domain.shared.PromotionCode

val A_MOMENT: Instant = Instant.parse("2026-09-09T10:00:00Z")

class CartBuilder {

    private val lines = mutableListOf<CartLine>()
    private var customerId: String? = null
    private var promotion: AppliedPromotion? = null

    fun holding(product: Product, quantity: Int = 1) = apply {
        lines.add(CartLine("line-${lines.size + 1}", product, quantity))
    }

    fun ownedBy(customerId: String) = apply { this.customerId = customerId }

    fun withPercentageOff(code: String, discountInCents: Int) = apply {
        promotion = AppliedPromotion(PromotionCode.of(code), PromotionKind.PERCENTAGE, Money.euro(discountInCents))
    }

    fun withFreeShipping(code: String) = apply {
        promotion = AppliedPromotion(PromotionCode.of(code), PromotionKind.FREE_SHIPPING, Money.NOTHING)
    }

    fun build() = Cart("cart-01", customerId, lines.toList(), promotion?.code, promotion, A_MOMENT)
}

fun aCart() = CartBuilder()
